File: php/lumen/app/Services/WhatsAppService.php
<?php

/**
 * خدمة إرسال رسائل واتساب عبر 4Jawaly API
 * WhatsApp messaging service via 4Jawaly API
 * 4Jawaly API کے ذریعے واٹس ایپ پیغامات بھیجنے کی سروس
 */

namespace App\Services;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;

class WhatsAppService
{
    /** @var Client */
    private Client $client;

    /** @var string عنوان أساسي للـ API - Base API URL - API کا بنیادی URL */
    private string $baseUrl;

    /** @var string بيانات المصادقة - Auth credentials - تصدیق کی معلومات */
    private string $authHeader;

    /**
     * إنشاء الخدمة - Constructor - کنسٹرکٹر
     */
    public function __construct()
    {
        $appKey = env('APP_KEY', '');
        $apiSecret = env('API_SECRET', '');
        $projectId = env('PROJECT_ID', '');

        $this->baseUrl = "https://api-users.4jawaly.com/api/v1/whatsapp/{$projectId}";
        $this->authHeader = 'Basic ' . base64_encode("{$appKey}:{$apiSecret}");

        $this->client = new Client([
            'base_uri' => $this->baseUrl,
            'timeout' => 30,
            'headers' => [
                'Content-Type' => 'application/json',
                'Authorization' => $this->authHeader,
            ],
        ]);
    }

    /**
     * إرسال طلب للـ API - Make API request - API درخواست بھیجیں
     *
     * @param array $messageData بيانات الرسالة - Message data - پیغام کا ڈیٹا
     * @return array استجابة JSON - JSON response - JSON جواب
     */
    private function sendRequest(array $messageData): array
    {
        $payload = [
            'path' => 'global',
            'params' => [
                'url' => 'messages',
                'method' => 'post',
                'data' => array_merge(
                    [
                        'messaging_product' => 'whatsapp',
                        'to' => $messageData['to'] ?? '',
                    ],
                    $messageData['payload'] ?? []
                ),
            ],
        ];

        try {
            $response = $this->client->post('', [
                'json' => $payload,
            ]);
            $body = (string) $response->getBody();
            return json_decode($body, true) ?? ['raw' => $body];
        } catch (GuzzleException $e) {
            return [
                'error' => true,
                'message' => $e->getMessage(),
                'code' => $e->getCode(),
            ];
        }
    }

    /**
     * إرسال رسالة نصية - Send text message - ٹیکسٹ پیغام بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param string $body نص الرسالة - Message body - پیغام کا متن
     * @return array
     */
    public function sendText(string $recipient, string $body): array
    {
        return $this->sendRequest([
            'to' => $recipient,
            'payload' => [
                'type' => 'text',
                'text' => ['body' => $body],
            ],
        ]);
    }

    /**
     * إرسال رسالة بأزرار تفاعلية - Send interactive buttons - انٹرایکٹو بٹن بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param string $bodyText نص الجسم - Body text - جسم کا متن
     * @param array $buttons مصفوفة الأزرار - Buttons array - بٹنوں کا سرے
     * @return array
     */
    public function sendButtons(string $recipient, string $bodyText, array $buttons): array
    {
        $formattedButtons = [];
        foreach ($buttons as $btn) {
            $formattedButtons[] = [
                'type' => 'reply',
                'reply' => [
                    'id' => $btn['id'] ?? '',
                    'title' => $btn['title'] ?? '',
                ],
            ];
        }

        return $this->sendRequest([
            'to' => $recipient,
            'payload' => [
                'type' => 'interactive',
                'interactive' => [
                    'type' => 'button',
                    'body' => ['text' => $bodyText],
                    'action' => ['buttons' => $formattedButtons],
                ],
            ],
        ]);
    }

    /**
     * إرسال رسالة بقائمة تفاعلية - Send interactive list - انٹرایکٹو لسٹ بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param string $headerText نص الترويسة - Header text - ہیڈر کا متن
     * @param string $bodyText نص الجسم - Body text - جسم کا متن
     * @param string $footerText نص التذييل - Footer text - فوٹر کا متن
     * @param string $buttonLabel تسمية زر القائمة - List button label - لسٹ بٹن لیبل
     * @param array $sections أقسام القائمة - List sections - لسٹ کے سیکشنز
     * @return array
     */
    public function sendList(
        string $recipient,
        string $headerText,
        string $bodyText,
        string $footerText,
        string $buttonLabel,
        array $sections
    ): array {
        return $this->sendRequest([
            'to' => $recipient,
            'payload' => [
                'type' => 'interactive',
                'interactive' => [
                    'type' => 'list',
                    'header' => ['type' => 'text', 'text' => $headerText],
                    'body' => ['text' => $bodyText],
                    'footer' => ['text' => $footerText],
                    'action' => [
                        'button' => $buttonLabel,
                        'sections' => $sections,
                    ],
                ],
            ],
        ]);
    }

    /**
     * إرسال صورة - Send image - تصویر بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param string $link رابط الصورة - Image link - تصویر کا لنک
     * @param string|null $caption وصف الصورة - Image caption - تصویر کا کیپشن
     * @return array
     */
    public function sendImage(string $recipient, string $link, ?string $caption = null): array
    {
        $image = ['link' => $link];
        if ($caption !== null) {
            $image['caption'] = $caption;
        }
        return $this->sendRequest([
            'to' => $recipient,
            'payload' => [
                'type' => 'image',
                'image' => $image,
            ],
        ]);
    }

    /**
     * إرسال فيديو - Send video - ویڈیو بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param string $link رابط الفيديو - Video link - ویڈیو کا لنک
     * @param string|null $caption وصف الفيديو - Video caption - ویڈیو کا کیپشن
     * @return array
     */
    public function sendVideo(string $recipient, string $link, ?string $caption = null): array
    {
        $video = ['link' => $link];
        if ($caption !== null) {
            $video['caption'] = $caption;
        }
        return $this->sendRequest([
            'to' => $recipient,
            'payload' => [
                'type' => 'video',
                'video' => $video,
            ],
        ]);
    }

    /**
     * إرسال ملف صوتي - Send audio - آڈیو بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param string $link رابط الملف الصوتي - Audio link - آڈیو کا لنک
     * @return array
     */
    public function sendAudio(string $recipient, string $link): array
    {
        return $this->sendRequest([
            'to' => $recipient,
            'payload' => [
                'type' => 'audio',
                'audio' => ['link' => $link],
            ],
        ]);
    }

    /**
     * إرسال مستند - Send document - دستاویز بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param string $link رابط المستند - Document link - دستاویز کا لنک
     * @param string|null $caption وصف المستند - Document caption - دستاویز کا کیپشن
     * @param string|null $filename اسم الملف - Filename - فائل کا نام
     * @return array
     */
    public function sendDocument(
        string $recipient,
        string $link,
        ?string $caption = null,
        ?string $filename = null
    ): array {
        $document = ['link' => $link];
        if ($caption !== null) {
            $document['caption'] = $caption;
        }
        if ($filename !== null) {
            $document['filename'] = $filename;
        }
        return $this->sendRequest([
            'to' => $recipient,
            'payload' => [
                'type' => 'document',
                'document' => $document,
            ],
        ]);
    }

    /**
     * إرسال طلب مسار مخصص - Send custom path request - مخصوص path درخواست بھیجیں
     * يستخدمه الموقع وجهة الاتصال (هيكل مختلف عن global)
     * Used by location and contact (different structure than global)
     *
     * @param string $path مسار الـ API - API path - API path
     * @param array $params المعاملات - Parameters - پیرامیٹرز
     * @return array
     */
    private function sendCustomPathRequest(string $path, array $params): array
    {
        $payload = [
            'path' => $path,
            'params' => $params,
        ];

        try {
            $response = $this->client->post('', [
                'json' => $payload,
            ]);
            $body = (string) $response->getBody();
            return json_decode($body, true) ?? ['raw' => $body];
        } catch (GuzzleException $e) {
            return [
                'error' => true,
                'message' => $e->getMessage(),
                'code' => $e->getCode(),
            ];
        }
    }

    /**
     * إرسال موقع جغرافي - Send location - مقام بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param float $lat خط العرض - Latitude - عرض البلد
     * @param float $lng خط الطول - Longitude - طول البلد
     * @param string $address العنوان - Address - پتہ
     * @param string|null $name اسم الموقع - Location name - مقام کا نام
     * @return array
     */
    public function sendLocation(
        string $recipient,
        float $lat,
        float $lng,
        string $address,
        ?string $name = null
    ): array {
        $params = [
            'phone' => $recipient,
            'lat' => $lat,
            'lng' => $lng,
            'address' => $address,
        ];
        if ($name !== null) {
            $params['name'] = $name;
        }
        return $this->sendCustomPathRequest('message/location', $params);
    }

    /**
     * إرسال جهة اتصال - Send contact - رابطہ بھیجیں
     *
     * @param string $recipient رقم المستلم - Recipient number - وصول کنندہ کا نمبر
     * @param array $contacts مصفوفة جهات الاتصال - Contacts array - رابطے کا سرے
     * @return array
     */
    public function sendContact(string $recipient, array $contacts): array
    {
        return $this->sendCustomPathRequest('message/contact', [
            'phone' => $recipient,
            'contacts' => $contacts,
        ]);
    }
}

File: flutter/whatsapp_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// WhatsApp Service for 4Jawaly API
/// خدمة واتساب لواجهة 4جوالي
/// 4Jawaly API کے لیے WhatsApp سروس
class WhatsAppService {
  /// App Key - مفتاح التطبيق - ایپ کی
  final String appKey;

  /// API Secret - سر التطبيق - API سیکرٹ
  final String apiSecret;

  /// Project ID - معرف المشروع - پروجیکٹ آئی ڈی
  final String projectId;

  /// Base URL for API - الرابط الأساسي للواجهة - API کا بنیادی URL
  static const String _baseUrl =
      'https://api-users.4jawaly.com/api/v1/whatsapp';

  /// Constructor - المُنشئ - کنسٹرکٹر
  /// [appKey] Application key - مفتاح التطبيق - ایپ کی
  /// [apiSecret] API secret - سر التطبيق - API سیکرٹ
  /// [projectId] Project identifier - معرف المشروع - پروجیکٹ آئی ڈی
  WhatsAppService({
    required this.appKey,
    required this.apiSecret,
    required this.projectId,
  });

  /// Make custom path API request - طلب API مع مسار مخصص - کسٹم پاتھ API درخواست
  /// Used for location/contact - different structure (path + direct params)
  /// [path] API path (e.g. message/location, message/contact) - المسار - API پاتھ
  /// [params] Direct params object - المعاملات المباشرة - براہ راست پیرامیٹرز
  Future<Map<String, dynamic>> _makeCustomPathRequest(
    String path,
    Map<String, dynamic> params,
  ) async {
    try {
      final credentials = base64Encode(utf8.encode('$appKey:$apiSecret'));
      final authHeader = 'Basic $credentials';

      final body = {'path': path, 'params': params};

      final url = Uri.parse('$_baseUrl/$projectId');
      final response = await http.post(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': authHeader,
        },
        body: jsonEncode(body),
      );

      dynamic parsedBody;
      try {
        parsedBody = response.body.isEmpty
            ? <String, dynamic>{}
            : jsonDecode(response.body);
      } catch (_) {
        parsedBody = {'raw': response.body};
      }

      return {
        'statusCode': response.statusCode,
        'body': parsedBody,
      };
    } catch (e) {
      return {
        'error': true,
        'message': e.toString(),
      };
    }
  }

  /// Make authenticated API request - تنفيذ طلب API مصادق عليه - تصدیق شدہ API درخواست
  /// [data] Request payload - بيانات الطلب - درخواست کا پے لوڈ
  Future<Map<String, dynamic>> _makeRequest(Map<String, dynamic> data) async {
    try {
      // Create Basic Auth header - إنشاء رأس المصادقة الأساسية - Basic Auth ہیڈر بنائیں
      final credentials = base64Encode(utf8.encode('$appKey:$apiSecret'));
      final authHeader = 'Basic $credentials';

      // Build request body - بناء جسم الطلب - درخواست کا جسم بنائیں
      final body = {
        'path': 'global',
        'params': {
          'url': 'messages',
          'method': 'post',
          'data': data,
        },
      };

      final url = Uri.parse('$_baseUrl/$projectId');
      final response = await http.post(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': authHeader,
        },
        body: jsonEncode(body),
      );

      // Parse response body - تحليل الاستجابة - جواب کا تجزیہ
      dynamic parsedBody;
      try {
        parsedBody = response.body.isEmpty
            ? <String, dynamic>{}
            : jsonDecode(response.body);
      } catch (_) {
        parsedBody = {'raw': response.body};
      }

      return {
        'statusCode': response.statusCode,
        'body': parsedBody,
      };
    } catch (e) {
      return {
        'error': true,
        'message': e.toString(),
      };
    }
  }

  /// Send text message - إرسال رسالة نصية - ٹیکسٹ میسج بھیجیں
  /// [to] Recipient phone (e.g. 9665XXXXXXXX) - رقم المستلم - ٹیلی فون نمبر
  /// [text] Message body - نص الرسالة - میسج کا متن
  Future<Map<String, dynamic>> sendText(String to, String text) async {
    final data = {
      'messaging_product': 'whatsapp',
      'to': to,
      'type': 'text',
      'text': {'body': text},
    };
    return _makeRequest(data);
  }

  /// Send interactive buttons - إرسال أزرار تفاعلية - تفاعلی بٹن بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [bodyText] Button list body text - نص جسم القائمة - بٹن کی فہرست کا متن
  /// [buttons] List of {id, title} - قائمة الأزرار - بٹنوں کی فہرست
  Future<Map<String, dynamic>> sendButtons(
    String to,
    String bodyText,
    List<Map<String, String>> buttons,
  ) async {
    final buttonList = buttons
        .map((b) => {
              'type': 'reply',
              'reply': {
                'id': b['id'] ?? '',
                'title': b['title'] ?? '',
              },
            })
        .toList();
    final data = {
      'messaging_product': 'whatsapp',
      'to': to,
      'type': 'interactive',
      'interactive': {
        'type': 'button',
        'body': {'text': bodyText},
        'action': {'buttons': buttonList},
      },
    };
    return _makeRequest(data);
  }

  /// Send interactive list - إرسال قائمة تفاعلية - تفاعلی فہرست بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [headerText] List header - عنوان القائمة - فہرست کا ہیڈر
  /// [bodyText] List body - نص جسم القائمة - فہرست کا متن
  /// [footerText] List footer - تذييل القائمة - فہرست کا فوٹر
  /// [buttonText] Action button text - نص زر العرض - بٹن کا متن
  /// [sections] List sections - أقسام القائمة - فہرست کے حصے
  Future<Map<String, dynamic>> sendList(
    String to, {
    required String headerText,
    required String bodyText,
    String? footerText,
    required String buttonText,
    required List<Map<String, dynamic>> sections,
  }) async {
    final data = {
      'messaging_product': 'whatsapp',
      'to': to,
      'type': 'interactive',
      'interactive': {
        'type': 'list',
        'header': {'type': 'text', 'text': headerText},
        'body': {'text': bodyText},
        if (footerText != null) 'footer': {'text': footerText},
        'action': {
          'button': buttonText,
          'sections': sections,
        },
      },
    };
    return _makeRequest(data);
  }

  /// Send image - إرسال صورة - تصویر بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [link] Image URL - رابط الصورة - تصویر کا URL
  /// [caption] Optional caption - وصف اختياري - اختیاری کیپشن
  Future<Map<String, dynamic>> sendImage(
    String to,
    String link, {
    String? caption,
  }) async {
    final data = {
      'messaging_product': 'whatsapp',
      'to': to,
      'type': 'image',
      'image': {
        'link': link,
        if (caption != null) 'caption': caption,
      },
    };
    return _makeRequest(data);
  }

  /// Send video - إرسال فيديو - ویڈیو بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [link] Video URL - رابط الفيديو - ویڈیو کا URL
  /// [caption] Optional caption - وصف اختياري - اختیاری کیپشن
  Future<Map<String, dynamic>> sendVideo(
    String to,
    String link, {
    String? caption,
  }) async {
    final data = {
      'messaging_product': 'whatsapp',
      'to': to,
      'type': 'video',
      'video': {
        'link': link,
        if (caption != null) 'caption': caption,
      },
    };
    return _makeRequest(data);
  }

  /// Send audio - إرسال ملف صوتي - آڈیو بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [link] Audio URL - رابط الملف الصوتي - آڈیو کا URL
  Future<Map<String, dynamic>> sendAudio(String to, String link) async {
    final data = {
      'messaging_product': 'whatsapp',
      'to': to,
      'type': 'audio',
      'audio': {'link': link},
    };
    return _makeRequest(data);
  }

  /// Send document - إرسال مستند - دستاویز بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [link] Document URL - رابط المستند - دستاویز کا URL
  /// [caption] Optional caption - وصف اختياري - اختیاری کیپشن
  /// [filename] Optional filename - اسم الملف - اختیاری فائل نام
  Future<Map<String, dynamic>> sendDocument(
    String to,
    String link, {
    String? caption,
    String? filename,
  }) async {
    final data = {
      'messaging_product': 'whatsapp',
      'to': to,
      'type': 'document',
      'document': {
        'link': link,
        if (caption != null) 'caption': caption,
        if (filename != null) 'filename': filename,
      },
    };
    return _makeRequest(data);
  }

  /// Send location - إرسال موقع جغرافي - مقام بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [lat] Latitude - خط العرض - عرض البلد
  /// [lng] Longitude - خط الطول - طول البلد
  /// [address] Optional address - العنوان - پتہ (اختیاری)
  /// [name] Optional name - الاسم - نام (اختیاری)
  Future<Map<String, dynamic>> sendLocation(
    String to,
    double lat,
    double lng, {
    String? address,
    String? name,
  }) async {
    final params = <String, dynamic>{
      'phone': to,
      'lat': lat,
      'lng': lng,
    };
    if (address != null) params['address'] = address;
    if (name != null) params['name'] = name;
    return _makeCustomPathRequest('message/location', params);
  }

  /// Send contact - إرسال جهة اتصال - رابطہ بھیجیں
  /// [to] Recipient phone - رقم المستلم - ٹیلی فون نمبر
  /// [contacts] List of contact objects - قائمة جهات الاتصال - رابطوں کی فہرست
  Future<Map<String, dynamic>> sendContact(
    String to,
    List<Map<String, dynamic>> contacts,
  ) async {
    final params = {
      'phone': to,
      'contacts': contacts,
    };
    return _makeCustomPathRequest('message/contact', params);
  }
}
